### PR TITLE
[TOREVIEW] Fb server added structure

### PR DESCRIPTION
**Pull request by LeonardBesseau**
Structure is implemented.
## Model
- Authentification is implemented with JsonWebToken.
- BaseController to try to avoid duplicating code for db request.
- ErrorController for middleware managing error.
- UserController for managing user specific behaviour.
## Model
Database model for user. Manages verification and some utility function concerning the model.
## Routes
- API route with authentification management

## Utils
- Utility features to manage incoming data
- Manages error

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+const xss = require('xss-clean');
+const hpp = require('hpp');
+const cors = require('cors');
+
+
+const userRoutes = require('./routes/userRoutes');
+const globalErrHandler = require('./controllers/errorController');
+const AppError = require('./utils/appError');
+const app = express();
+
```
Comment (LeonardBesseau) [inline app.js:15]: Add middleware to manage behaviour we want but don't want to manage ourself

File: .gitignore [hunk 2]
```diff
@@ -104,3 +104,8 @@ dist
 .tern-port
 
 ./idea
```
Comment (K-do) [inline .gitignore:106]: il sert à rien, vuq ue c'est un .idea, tu peux le supprimer

**Review by K-do — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (LeonardBesseau) [reply .gitignore:106]: Auto added by npm

File: .gitignore [hunk 3]
```diff
@@ -104,3 +104,8 @@ dist
...
+.idea/
+.yarn/
+package-lock.json
+yarn.lock
+.yarn/
```
[the remark this answers is not in the document]
Comment (LeonardBesseau) [reply .gitignore]: Agreed

File: controllers/authController.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        next(err);
+    }
+};
+
+// Authorization check if the user have rights to do this action
+exports.restrictTo = (...roles) => {
+    return (req, res, next) => {
+        if (!roles.includes(req.user.role)) {
+            return next(
+                new AppError(403, "fail", "You are not allowed to do this action"),
```
[the remark this answers is not in the document]
Comment (LeonardBesseau) [reply controllers/authController.js:138]: This not an exception. This is an error. An unauthorized access is recoverable and we want to return to the user making the request that something went wrong

File: controllers/authController.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+const AppError = require("../utils/appError");
+
+const createToken = id => {
+    return jwt.sign(
+        {
+            id,
+        },
+        process.env.JWT_SECRET,
+        {
+            expiresIn: process.env.JWT_EXPIRES_IN,
```
Comment (K-do) [inline controllers/authController.js:13]: algorithm: ? (ECDSA, HMAC ?)
Do we use JWT for encryption or signing ?

Comment (LeonardBesseau) [reply controllers/authController.js:13]: https://jwt.io/introduction/
Digital signature. We can choose to use 

   - HMAC + SHA256
   - RSASSA-PKCS1-v1_5 + SHA256
   - ECDSA + P-256 + SHA256
   
   Normally, the defaul is HMAC

File: config.env [hunk 6]
```diff
@@ -0,0 +1,11 @@
+NODE_ENVIROMMENT=development
+PORT=3000
+DATABASE=mongodb://${MONGO_USERNAME}:${MONGO_PASSWORD}@${MONGO_HOSTNAME}:${MONGO_PORT}/${MONGO_DB}?authSource=admin
+MONGO_USERNAME= 
+MONGO_PASSWORD =
```
[the remark this answers is not in the document]
Comment (LeonardBesseau) [reply config.env:5]: This is a public repository

[the remark this answers is not in the document]
Comment (LeonardBesseau) [reply config.env:5]: I'll let you work the rest yourself.
This file is here to have the structure correct. Nothing else

[the remark this answers is not in the document]
Comment (K-do) [reply config.env:5]: good point

File: controllers/authController.js (hunk 5 above)
Comment (K-do) [reply controllers/authController.js:13]: Do we add "role" in the payload too ?

File: controllers/authController.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+exports.signup = async (req, res, next) => {
+    try {
+        const user = await User.create({
+            name: req.body.name,
+            email: req.body.email,
+            password: req.body.password,
+            role: req.body.role,
+        });
+
+        const token = createToken(user.id);
```
Comment (K-do) [inline controllers/authController.js:73]: I suppose the token is here, because we sign in just after creation ?

File: utils/appError.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class AppError extends Error {
+    constructor(statusCode, status, message) {
+        super(message);
+        this.statusCode = statusCode;
+        this.status = status;
+        this.message = message;
```
Comment (K-do) [inline utils/appError.js:6]: duplicate with 3 ?

**Review by K-do — CHANGES_REQUESTED** [review]
Where are the tests

**Comment by LeonardBesseau** [comment]
> Where are the tests ?

I ate them

File: controllers/authController.js (hunk 5 above)
Comment (LeonardBesseau) [reply controllers/authController.js:13]: Will do

File: utils/appError.js (hunk 8 above)
Comment (LeonardBesseau) [reply utils/appError.js:6]: No. After check the message passed in the constructor is to add the message to the stack trace and avoid a duplication of the message. the this message is to have a visible error message to extract easily to later display to the user

Comment (LeonardBesseau) [reply utils/appError.js:6]: image
Expected result

**Comment by LeonardBesseau** [comment]
Modified code as requested.
Tests are on the way

File: models/eventModel.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        validate: [validator.isIP, "IP is invalid"],
+        immutable: true
+    },
+    userAgent: {
+        type: String,
+        required: true,
+        immutable: true
+    },
+    user: {
+        type: mongoose.Schema.Types.ObjectId,
```
Comment (K-do) [inline models/eventModel.js:26]: needs ref: ?

File: models/eventModel.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        required: true,
+        immutable: true
+    }
+});
+
+const eventOrder = new mongoose.Schema({
+    user: {
+        type: mongoose.Schema.Types.ObjectId,
+        required: true,
```
Comment (K-do) [inline models/eventModel.js:35]: same

File: models/eventModel.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        immutable: true
+    },
+    order: {
+        type: mongoose.Schema.Types.ObjectId,
+        required: true,
```
Comment (K-do) [inline models/eventModel.js:40]: same

File: models/eventModel.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        immutable: true
+    }
+});
+
+const eventElement = new mongoose.Schema({
```
Comment (K-do) [inline models/eventModel.js:45]: same

File: routes/eventRoutes.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+router
+    .route('/connections')
+    .get(eventController.getAllConnectionEvent);
+
+router
+    .route('/connections/:id')
+    .get(eventController.getConnectionEvent);
+
+router
+    .route('/connections/user/:id')
```
Comment (K-do) [inline routes/eventRoutes.js:21]: you could have : /user/:userId/:id like I did for Element if u think is a good idea

File: controllers/eventController.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+const {ConnectionEvent, OrderEvent, ElementEvent} = require('../models/eventModel');
+const base = require('./baseController');
+
+/**
+ * Util function to get an event based on a criteria
+ * @param Model the model to make the request from
+ * @param param criteria of the search
+ * @returns {function(req, res, next): Promise<Document[]>}
+ */
+const getEvent = (Model, param) => async ( req, res, next) => {
```
Comment (K-do) [inline controllers/eventController.js:11]: I choose a different way with getAllElementsByLocation in elementController

File: tests/models/eventModelTest.js [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+            it('should throws when value is not a correct ip', async () => {
+                await expect(ConnectionEvent.create({
+                    ip: "127.0.",
+                    userAgent: "Red Star OS 4",
+                    user: "606afdb5aa09d43a84b6181a"
+                })).to.be.rejectedWith(Error);
+            });
+        });
+        describe('user', function () {
+            it('should throws when value is not a correct id', async () => {
```
Comment (K-do) [inline tests/models/eventModelTest.js:68]: wich id ? you mean user ?

File: tests/models/userModelTest.js [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            await expect(User.create({
+                firstName: "test",
+                lastName: "test",
+                email: "email@email.test",
+                password: "012345678",
+                role: "NOPE"
+            })).to.be.rejectedWith(Error);
+        });
+    });
+    describe('Role', function () {
```
Comment (K-do) [inline tests/models/userModelTest.js:69]: duplicate with line 58

**Review by K-do — COMMENTED** [review]
I like it, but still have some questions

File: tests/models/eventModelTest.js (hunk 15 above)
Comment (LeonardBesseau) [reply tests/models/eventModelTest.js:68]: Ref does not check if it is a valid object id for the reference (Require another query). Ref is used with populate to automatically fetch the associated data.

File: models/eventModel.js (hunk 9 above)
Comment (LeonardBesseau) [reply models/eventModel.js:26]: Do we need to populate ?. I will add them anyway but still something we need to see

File: models/eventModel.js (hunk 10 above)
Comment (LeonardBesseau) [reply models/eventModel.js:35]: same as before

File: models/eventModel.js (hunk 11 above)
Comment (LeonardBesseau) [reply models/eventModel.js:40]: same as before

File: models/eventModel.js (hunk 12 above)
Comment (LeonardBesseau) [reply models/eventModel.js:45]: same as before

File: routes/eventRoutes.js (hunk 13 above)
Comment (LeonardBesseau) [reply routes/eventRoutes.js:21]: Didi not see what you are talking about in your pr. Will discuss later

File: tests/models/userModelTest.js (hunk 16 above)
Comment (LeonardBesseau) [reply tests/models/userModelTest.js:69]: Yes. Was a bit to liberal with my copy-paste

File: models/eventModel.js (hunk 9 above)
Comment (K-do) [reply models/eventModel.js:26]: 👍🏼

**Review by K-do — APPROVED** [review]

**Pull request merged**